hail/src/main/scala/is/hail/expr/ir/ForwardLets.scala

```diff
@@ -0,0 +1,97 @@
+package is.hail.expr.ir
+
+import is.hail.utils._
+
+import scala.collection.mutable
+
+object ForwardLets {
+  def apply(ir0: BaseIR): BaseIR = {
+    val ir1 = new NormalizeNames(allowFreeVariables = true).apply(ir0)
+    val UsesAndDefs(uses, _) = ComputeUsesAndDefs(ir1, freeVariablesError = false)
+    val nestingDepth = NestingDepth(ir1)
+
+    def rewriteTable(tir: TableIR): BaseIR = tir.copy(tir
+      .children
+      .iterator
+      .zipWithIndex
+      .map {
+      case (child: IR, i) => rewrite(child, NewBindings(tir, i).dropBindings)
+      case (tir: TableIR, _) => rewriteTable(tir)
+      case (mir: MatrixIR, _) => rewriteMatrix(mir)
+      case (bmir: BlockMatrixIR, _) => rewriteBlockMatrix(bmir)
+    }.toFastIndexedSeq)
+
+    def rewriteMatrix(mir: MatrixIR): BaseIR = mir.copy(mir
+      .children
+      .iterator
+      .zipWithIndex
+      .map {
+        case (child: IR, i) => rewrite(child, NewBindings(mir, i).dropBindings)
+        case (tir: TableIR, _) => rewriteTable(tir)
+        case (mir: MatrixIR, _) => rewriteMatrix(mir)
+        case (bmir: BlockMatrixIR, _) => rewriteBlockMatrix(bmir)
+      }.toFastIndexedSeq)
+
+    def rewriteBlockMatrix(bmir: BlockMatrixIR): BaseIR = bmir.copy(bmir
+      .children
+      .iterator
+      .zipWithIndex
+      .map {
+        case (child: IR, i) => rewrite(child, NewBindings(bmir, i).dropBindings)
+        case (tir: TableIR, _) => rewriteTable(tir)
+        case (mir: MatrixIR, _) => rewriteMatrix(mir)
+        case (bmir: BlockMatrixIR, _) => rewriteBlockMatrix(bmir)
+      }.toFastIndexedSeq)
+
+    def rewrite(ir: IR, env: BindingEnv[IR]): IR = {
+
+      def shouldForward(value: IR, refs: mutable.Set[RefEquality[Ref]], base: IR): Boolean = {
+        value.isInstanceOf[Ref] ||
+          IsConstant(value) ||
```

Comment:
I wonder, is it a good idea to always forward large literals and strings?

Reply:
Right now, yes. Literals are lifted out in a separate pass that deduplicates, so this will generate cleaner IRs.

Do we deduplicate `Str` literals in the emitter? We should.